metadata-ingestion/src/datahub/utilities/file_backed_collections.py

```diff
@@ -131,15 +208,102 @@ def __iter__(self) -> Iterator[str]:
             (*self._active_object_cache.keys(),),
         )
         row = cursor.fetchone()
 
         return row[0] + len(self._active_object_cache)
 
+    def sql_query(
+        self,
+        query: str,
+        params: Tuple[Any, ...] = (),
+        refs: Optional[List[Union["FileBackedList", "FileBackedDict"]]] = None,
+    ) -> List[Tuple[Any, ...]]:
+        # We need to flush object and any objects the query references to ensure
+        # that we don't miss objects that have been modified but not yet flushed.
+        self.flush()
+        if refs is not None:
+            for referenced_table in refs:
+                referenced_table.flush()
+
+        cursor = self._conn.execute(query, params)
+        return cursor.fetchall()
+
     def close(self) -> None:
-        self._conn.close()
+        _sqlite_connection_cache.drop_connection(self.filename)
 
     def __del__(self) -> None:
         self.close()
+
+
+class FileBackedList(Generic[_VT]):
+    """
+    An append-only, list-like object that stores its contents in a SQLite database.
+
+    This class is not thread-safe.
+    """
+
+    _len: int = field(default=0)
+    _dict: FileBackedDict[_VT] = field(init=False)
+
+    def __init__(
+        self,
+        filename: pathlib.Path,
+        serializer: Callable[[_VT], SqliteValue],
+        deserializer: Callable[[Any], _VT],
+        tablename: str = _DEFAULT_TABLE_NAME,
+        extra_columns: Optional[Dict[str, Callable[[_VT], SqliteValue]]] = None,
+        cache_max_size: Optional[int] = None,
+        cache_eviction_batch_size: Optional[int] = None,
+    ) -> None:
+        self._len = 0
+        self._dict = FileBackedDict(
+            filename=filename,
+            serializer=serializer,
+            deserializer=deserializer,
+            tablename=tablename,
+            extra_columns=extra_columns or {},
+            cache_max_size=cache_max_size or _DEFAULT_MEMORY_CACHE_MAX_SIZE,
+            cache_eviction_batch_size=cache_eviction_batch_size
+            or _DEFAULT_MEMORY_CACHE_EVICTION_BATCH_SIZE,
+        )
+
+    @property
+    def tablename(self) -> str:
+        return self._dict.tablename
+
+    def __getitem__(self, index: int) -> _VT:
+        if index < 0 or index >= self._len:
+            raise IndexError(f"list index {index} out of range")
+
+        return self._dict[str(index)]
+
+    def __setitem__(self, index: int, value: _VT) -> None:
+        if index < 0 or index >= self._len:
+            raise IndexError(f"list index {index} out of range")
+
+        self._dict[str(index)] = value
+
+    def append(self, value: _VT) -> None:
+        self._dict[str(self._len)] = value
+        self._len += 1
+
+    def __len__(self) -> int:
+        return self._len
+
+    def __iter__(self) -> Iterator[_VT]:
+        for index in range(self._len):
+            yield self[index]
+
+    def flush(self) -> None:
+        self._dict.flush()
+
+    def sql_query(
```

Comment:
I'd consider an additional abstraction called a catalog which can be created with a list of "base tables" that you can issue queries against. Just a suggestion.

A bit weird to have the method to query the thing on the thing itself